---
title: "Shiny App Phosphosites"
author: "Andri Widmer"
date: "`r Sys.Date()`"
output: html_document
---

```{r}

#Use this snippet to start launch the created site (shiny). It allows you to explore phosphosites of Groups 1,2,3 and 13. More data will follow...

library(tableHTML)
library(htmltools)
library(shiny)
library(bslib)
library(xml2)

# Load data from GitHub
MSA_as_df_git=readRDS(url("https://raw.githubusercontent.com/WidmerAndri/msa_files/main/MSA_as_df_git.rds"))

# Use only the first 3 groups and 13 for display
msa_list = MSA_as_df_git[c(1:3,15)]
names(msa_list) = paste0("Group ", c(1:3,15))

# Extract real group numbers for correct linking
numbers = sapply(msa_list, function(df) df$Group_Nr[1])
fasta_links = paste0("https://raw.githubusercontent.com/WidmerAndri/msa_files/refs/heads/main/Modified_MSA_group_ClustalO_", numbers, ".fasta")

# UI layout
ui = fluidPage(
  titlePanel("Phosphosites in Orthologous Proteins of Yeast, Arabidopsis and Human"),
  
  HTML("
  <p>This tool allows you to explore multiple sequence alignments (MSAs) and associated phosphosite annotations across <b>Homo sapiens</b>, <b>Saccharomyces cerevisiae</b>, and <b>Arabidopsis thaliana</b>.</p>
  <p>The phosphorylation information comes from EPSD and was downloaded on February 26, 2025.</p>
  <p><b>Open in Jalview</b> to display the MSA (may take a few seconds). Click on a UniProt ID to view protein details.</p>
  <p>Capitalized letters represent known phosphorylated sites. The guiding sequence uses <code>-</code> and <code>^</code> to show conserved phosphosites. Use Jalview's colour scheme <i>(Colour → By Annotation → Conservation)</i> for better insight.</p>
  <p>For easier identification of phosphosites, press <b>Open in Viewer</b>. The greener, the higher the score.</p>
"),

  br(),
  
  selectInput("msa_choice", "Select Group:", choices = names(msa_list)),
  uiOutput("jalview_link"),
  br(),
  
  uiOutput("group_title"),
  htmlOutput("msa_table")
)

# Server logic
server = function(input, output, session) {
  
  output$group_title = renderUI({
    req(input$msa_choice)
    index = which(names(msa_list) == input$msa_choice)
    h4(paste0("Information for Group ", numbers[index]))
  })
  
output$msa_table = renderUI({
  req(input$msa_choice)
  df = msa_list[[input$msa_choice]]
  
  # Only first 4 columns
  df = df[, 1:4]
  
  # Make UniProt ID in first column clickable
  df[[1]] = paste0('<a href="https://www.uniprot.org/uniprotkb/', df[[1]], '/entry" target="_blank">', df[[1]], '</a>')
  
  # Inline color styling in 4th column (cell by cell)
  df[[4]] = ifelse(df[[4]] == "P-Sites from EPSD",
                   paste0('<div style="background-color: lightgreen;">', df[[4]], '</div>'),
                   paste0('<div style="background-color: lightcoral;">', df[[4]], '</div>'))
  
  # Build HTML table
  tableHTML(df,
            widths = c(200, 100, 300, 150),
            rownames = FALSE,
            escape = FALSE)
})

  output$jalview_link = renderUI({
    req(input$msa_choice)
    index = which(names(msa_list) == input$msa_choice)
    # "Group 1" → 1
    group_label = input$msa_choice
    group_nr = as.numeric(gsub("Group ", "", group_label))
    fasta_url=URLencode(paste0("https://raw.githubusercontent.com/WidmerAndri/msa_files/refs/heads/main/Modified_MSA_group_ClustalO_", group_nr, ".fasta"))
    
    tagList(
    tags$a("Open in JalviewJS",
           href = paste0("https://jalview.github.io/jalview-js/JalviewJS.html?open%20", fasta_url),
           target = "_blank",
           style = "margin-right: 10px; padding: 6px 12px; background-color: #007bff; color: white; text-decoration: none; border-radius: 5px;"),
    
    tags$a("Open in Viewer",
           href = paste0("https://widmerandri.github.io/msa_files/MSA_htmls/Versuch", group_nr, ".html"),
           target = "_blank",
           style = "padding: 6px 12px; background-color: red; color: white; text-decoration: none; border-radius: 5px;")
  )
  })
}

shinyApp(ui, server)


rm(fasta_links,numbers)

```

